### PR TITLE
[LoopReduceMotion] Improve loop by extract reduction instruction

### PR DESCRIPTION
**Pull request by Anjian-Wen**
Improve loop by extract reduction instruction with vector_reduce_add.

This pass is designed to sink `ReduceCall` operations out of loops to reduce
the number of instructions within the loop body. Below are the target pattern to be matched and the resulting pattern
after the transformation. 

```text
before                   | after
--------------------------------------------------
loop:                    | loop:
  ...                    |   ...
  d = reduce_add vc      |   vsum = vadd vsum, vc
  sum = add sum, d       |   ...
  ...                    |   ...
exit:                    | exit:
  value = sum            |   d = reduce_add vsum
  ...                    |   value = d
  ...                    |   ...
```

### CODE REVIEW

**Comment by Anjian-Wen** [comment]
> Are the vector loops that this optimizes generated by the loop vectorizer? It looks like this is essentially converting an in-loop reduction to an out-of-loop reduction, but my understanding was that we did in-loop reductions by default anyway?

Thanks for the commit. In my case, It seems that the IR Dump After LoopVectorizePass haven't generated reduction, but generated by SLP. below are the code I try to optimize.

```
int pixel_asd8( pixel *pix1, intptr_t stride1, pixel *pix2, intptr_t stride2, int height )
{
    int sum = 0;
    for( int y = 0; y < height; y++, pix1 += stride1, pix2 += stride2 )
        for( int x = 0; x < 8; x++ )
            sum += pix1[x] - pix2[x];
    return abs( sum );
}
```

File: llvm/include/llvm/Transforms/Vectorize/LoopReduceMotion.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
+//===- LoopReduceMotion.h ---------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass is designed to hoist `ReduceCall` operations out of loops to reduce
```
Comment (topperc) [inline llvm/include/llvm/Transforms/Vectorize/LoopReduceMotion.h]: If you're moving it to the exit block, it's 'sink' not 'hoist'. That's how LICM refers to it.

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+      continue;
+    // pattern match success
+    LLVM_DEBUG(dbgs() << "FRM: Found pattern to optimize in loop "
+                      << Header->getName() << "!\n");
+
+    VectorType *VecTy = cast<VectorType>(VecBin->getType());
+    IRBuilder<> PreheaderBuilder(Preheader->getTerminator());
+
+    Value *VecZero = PreheaderBuilder.CreateVectorSplat(
+        VecTy->getElementCount(), ConstantInt::get(VecTy->getElementType(), 0),
```
Comment (topperc) [inline llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: ConstantInt::get an create a splat if you give it a vector type, you don't need to splat the a constant 0.

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        PreheaderValue
+            ? LandingPadBuilder.CreateAdd(PreheaderValue, ScalarTotalSum)
+            : ScalarTotalSum;
+
+    // replace use of phi and erase use empty value
+    if (!PN.use_empty())
+      PN.replaceAllUsesWith(PoisonValue::get(PN.getType()));
+    if (PN.use_empty())
+      PN.eraseFromParent();
+    RecurrenceInst->replaceAllUsesWith(dyn_cast<Instruction>(LastAdd));
```
Comment (topperc) [inline llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: Can this dyn_cast fail?

File: llvm/test/Transforms/LoopReduceMotion/loop-reduce-motion-test.ll [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+; CHECK-NEXT:    [[VEC_SUM_NEXT]] = add <8 x i32> [[VEC_SUM_PHI]], [[TMP4]]
+; CHECK-NEXT:    [[INC9]] = add nuw nsw i32 [[Y_025]], 1
+; CHECK-NEXT:    [[ADD_PTR]] = getelementptr inbounds i8, ptr [[PIX1_ADDR_023]], i64 [[STRIDE1:%.*]]
+; CHECK-NEXT:    [[ADD_PTR10]] = getelementptr inbounds i8, ptr [[PIX2_ADDR_022]], i64 [[STRIDE2:%.*]]
+; CHECK-NEXT:    [[EXITCOND_NOT:%.*]] = icmp eq i32 [[INC9]], [[HEIGHT]]
+; CHECK-NEXT:    br i1 [[EXITCOND_NOT]], label [[LOOP_EXIT_LANDING:%.*]], label [[FOR_COND1_PREHEADER]]
+; CHECK:       loop.exit.landing:
+; CHECK-NEXT:    [[SCALAR_TOTAL_SUM:%.*]] = call i32 @llvm.vector.reduce.add.v8i32(<8 x i32> [[VEC_SUM_NEXT]])
+; CHECK-NEXT:    [[TMP5:%.*]] = add i32 0, [[SCALAR_TOTAL_SUM]]
+; CHECK-NEXT:    br label [[FOR_COND_CLEANUP_LOOPEXIT:%.*]]
```
Comment (topperc) [inline llvm/test/Transforms/LoopReduceMotion/loop-reduce-motion-test.ll]: Can we use the existing exit block if it has only one predecessor instead of creating a new block?

**Comment by Anjian-Wen** [comment]
@topperc Thanks for your careful review, I will go through the comments and fix them one by one

File: llvm/lib/Transforms/Vectorize/CMakeLists.txt [hunk 5]
```diff
@@ -3,6 +3,7 @@ add_llvm_component_library(LLVMVectorize
   LoopIdiomVectorize.cpp
   LoopVectorizationLegality.cpp
   LoopVectorize.cpp
+  LoopReduceMotion.cpp
```
[the remark this answers is not in the document]
Comment (Anjian-Wen) [reply llvm/lib/Transforms/Vectorize/CMakeLists.txt]: Fixed!

File: llvm/test/Transforms/LoopReduceMotion/loop-reduce-motion-test.ll [hunk 6]
```diff
@@ -0,0 +1,68 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; loop-reduce-motion-test.ll
+; RUN: opt -passes=loop-reduce-motion -S < %s | FileCheck %s
+
+define i32 @pixel_asd8(ptr %pix1, i64 %stride1, ptr %pix2, i64 %stride2, i32 %height) {
```
Comment (fhahn) [inline llvm/test/Transforms/LoopReduceMotion/loop-reduce-motion-test.ll]: It looks like this is missing a bunch of test coverage, mostly negative tests, but would also be good to have multi block/multi exit loops.

**Review by fhahn — COMMENTED** [review]
I don't think here is anything RISCV specific about this pattch. I think it would be preferable if it would just run after the vectorizers generally?

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+        !(CalledFunc->getIntrinsicID() == Intrinsic::vector_reduce_add))
+      continue;
+
+    Value *ReduceOperand = ReduceCall->getArgOperand(0);
+    Instruction *VecBin = dyn_cast<Instruction>(ReduceOperand);
+    if (!VecBin || (VecBin->getOpcode() != Instruction::Sub &&
+                    VecBin->getOpcode() != Instruction::Add))
+      continue;
+    // pattern match success
+    LLVM_DEBUG(dbgs() << "FRM: Found pattern to optimize in loop "
```
[the remark this answers is not in the document]
Comment (Anjian-Wen) [reply llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: Thanks, I will remove it in case of misunderstanding.

File: llvm/test/Transforms/LoopReduceMotion/loop-reduce-motion-test.ll (hunk 4 above)
Comment (Anjian-Wen) [reply llvm/test/Transforms/LoopReduceMotion/loop-reduce-motion-test.ll]: Yes, from my case, the existing exit block get cleaned up after the `CodeGen Prepare Pass` if its not needed. but the new created branch which do the final reduce_sum and add will go through SelectionDAG. According to [fhahn](https://github.com/fhahn) suggestions, I will add more test to verify it

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+#include "llvm/Plugins/PassPlugin.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include "llvm/Transforms/Utils/Local.h"
+#include "llvm/Transforms/Utils/LoopUtils.h"
+
+#define DEBUG_TYPE "loop-reduce-motion"
+
+using namespace llvm;
+
+class LoopReduceMotion : public FunctionPass {
```
Comment (mshockwave) [inline llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: could you give this Pass a different name like "LoopReduceMotionLegacy", otherwise it's a bit easy to get confused with its new PassManager counterpart.

File: llvm/lib/Target/RISCV/RISCVTargetMachine.cpp [hunk 9]
```diff
@@ -479,6 +479,7 @@ void RISCVPassConfig::addIRPasses() {
     addPass(createRISCVGatherScatterLoweringPass());
     addPass(createInterleavedAccessPass());
     addPass(createRISCVCodeGenPrepareLegacyPass());
+    addPass(createLoopReduceMotionPass());
```
Comment (mshockwave) [inline llvm/lib/Target/RISCV/RISCVTargetMachine.cpp]: does it have to run in the codegen pipeline? can we run it at the end of middle-end optimization pipeline? If it has to, which codegen Pass created this pattern?

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+    Value *LastAdd =
+        PreheaderValue
+            ? LandingPadBuilder.CreateAdd(PreheaderValue, ScalarTotalSum)
+            : ScalarTotalSum;
+
+    // replace use of phi and erase use empty value
+    if (!PN.use_empty())
+      PN.replaceAllUsesWith(PoisonValue::get(PN.getType()));
+    if (PN.use_empty())
+      PN.eraseFromParent();
```
[the remark this answers is not in the document]
Comment (Anjian-Wen) [reply llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: good idea, fixed!

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+};
+
+char LoopReduceMotion::ID = 0;
+
+PreservedAnalyses LoopReduceMotionPass::run(Function &F,
+                                            FunctionAnalysisManager &FAM) {
+  LoopInfo &LI = FAM.getResult<LoopAnalysis>(F);
+  DominatorTree &DT = FAM.getResult<DominatorTreeAnalysis>(F);
+  bool Changed = false;
+  for (Loop *L : LI) {
```
[the remark this answers is not in the document]
Comment (lukel97) [reply llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: Sorry for the delay in the review, it looks like we don't add it to the RISCV pipeline specifically anymore, I wonder if it's still possible?

File: llvm/lib/Passes/PassBuilderPipelines.cpp [hunk 12]
```diff
@@ -1418,6 +1419,8 @@ void PassBuilder::addVectorPasses(OptimizationLevel Level,
   }
   // Enhance/cleanup vector code.
   FPM.addPass(VectorCombinePass());
+  // Try to sink ReduceCall out of loop
+  FPM.addPass(LoopReduceMotionPass());
```
Comment (lukel97) [inline llvm/lib/Passes/PassBuilderPipelines.cpp]: I haven't tried this but if you wanted to make this a loop pass it would probably look like:

```suggestion
    LoopPassManager LPM;
    LPM.addPass(LoopReductionMotionPass());
    FPM.addPass(
        createFunctionToLoopPassAdaptor(std::move(LPM)));
```

Comment (Anjian-Wen) [reply llvm/lib/Passes/PassBuilderPipelines.cpp]: Thanks, I'll give this way a try.

**Review by davemgreen — COMMENTED** [review]
I think this would be detrimental to any architectures that preferred inloop reductions or could perform the reduction as or more efficiently than the add. (I was trying to test it, but it crashed and looks like it is fairly limited to what it can transform at the moment).

This probably needs to be cost-modelled, although the cost-model can be difficult as there are so many operations potentially transformed into a single instruction.

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+                                 : RecurrenceInst->getOperand(0);
+
+    CallInst *ReduceCall = dyn_cast<CallInst>(RecurrenceValue);
+    if (!ReduceCall)
+      continue;
+    Function *CalledFunc = ReduceCall->getCalledFunction();
+
+    if (!CalledFunc || !CalledFunc->isIntrinsic() ||
+        !(CalledFunc->getIntrinsicID() == Intrinsic::vector_reduce_add))
+      continue;
```
[the remark this answers is not in the document]
Comment (Anjian-Wen) [reply llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: done!

File: llvm/test/Transforms/PhaseOrdering/AArch64/udotabd.ll [hunk 14]
```diff
@@ -9,361 +9,325 @@ define dso_local i32 @test(ptr noundef %p1, i32 noundef %s_p1, ptr noundef %p2,
...
-; CHECK-O3-NEXT:    [[IDX_EXT:%.*]] = sext i32 [[S_P1]] to i64
 ; CHECK-O3-NEXT:    [[TMP0:%.*]] = load <16 x i8>, ptr [[P1]], align 1, !tbaa [[CHAR_TBAA0:![0-9]+]]
 ; CHECK-O3-NEXT:    [[TMP1:%.*]] = zext <16 x i8> [[TMP0]] to <16 x i16>
 ; CHECK-O3-NEXT:    [[TMP2:%.*]] = load <16 x i8>, ptr [[P2]], align 1, !tbaa [[CHAR_TBAA0]]
 ; CHECK-O3-NEXT:    [[TMP3:%.*]] = zext <16 x i8> [[TMP2]] to <16 x i16>
 ; CHECK-O3-NEXT:    [[TMP4:%.*]] = sub nsw <16 x i16> [[TMP1]], [[TMP3]]
 ; CHECK-O3-NEXT:    [[TMP5:%.*]] = tail call <16 x i16> @llvm.abs.v16i16(<16 x i16> [[TMP4]], i1 false)
 ; CHECK-O3-NEXT:    [[TMP6:%.*]] = zext <16 x i16> [[TMP5]] to <16 x i32>
-; CHECK-O3-NEXT:    [[TMP7:%.*]] = tail call i32 @llvm.vector.reduce.add.v16i32(<16 x i32> [[TMP6]])
+; CHECK-O3-NEXT:    [[IDX_EXT:%.*]] = sext i32 [[S_P1]] to i64
```
[the remark this answers is not in the document]
Comment (Anjian-Wen) [reply llvm/test/Transforms/PhaseOrdering/AArch64/udotabd.ll]: Great suggestions，`preferInLoopReduction` is a good idea, I will try to enrich the costmodel with getExtendedReductionCost and getMulAccReductionCost and  more types of input‘s cost.

[the remark this answers is not in the document]
Comment (Anjian-Wen) [reply llvm/test/Transforms/PhaseOrdering/AArch64/udotabd.ll]: I added preferInLoopReduction, and add more CostModel patterns with getExtendedReductionCost and getMulAccReductionCost refer to some descriptions of ReductionCost in other Pass. Can you take a look at this version?

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+          InstructionCost ExtCost0 =
+              TTI.getCastInstrCost(Op0->getOpcode(), VectorTy, Op0SrcTy,
+                                   TTI::CastContextHint::None, CostKind, Op0);
+          InstructionCost ExtCost1 =
+              TTI.getCastInstrCost(Op1->getOpcode(), VectorTy, Op1SrcTy,
+                                   TTI::CastContextHint::None, CostKind, Op1);
+          InstructionCost MulCost =
+              TTI.getArithmeticInstrCost(Instruction::Mul, VectorTy, CostKind);
+          InstructionCost RedCost = TTI.getMulAccReductionCost(
+              IsUnsigned, RedOp->getOpcode(), Ty->getScalarType(),
```
[the remark this answers is not in the document]
Comment (Anjian-Wen) [reply llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: fixed！

**Comment by Anjian-Wen** [comment]
kindly ping to keep pr alive

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+      if (Instruction *Inst = dyn_cast<Instruction>(U)) {
+        BasicBlock *BB = Inst->getParent();
+        if (Inst != PN && L.contains(BB)) {
+          hasOtherUse = true;
+          break;
+        }
+      }
+    }
+    if (hasOtherUse)
+      continue;
```
Comment (arsenm) [inline llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: Avoid assign and break pattern, maybe use any_of?

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+            RecurrenceValue,
+            llvm::PatternMatch::m_Intrinsic<Intrinsic::vector_reduce_add>(
+                llvm::PatternMatch::m_Value(ReduceOperand))))
+      continue;
+
+    CallInst *ReduceCall = dyn_cast<CallInst>(RecurrenceValue);
+    Instruction *VecIn = dyn_cast<Instruction>(ReduceOperand);
+    VectorType *VecTy = cast<VectorType>(VecIn->getType());
+
+    auto TTI = &AR.TTI;
```
Comment (arsenm) [inline llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]:
```suggestion
    TargetTransformInfo &TTI = &AR.TTI;
```

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+            : ScalarTotalSum;
+    // replace the use of Recurrence Node and delete the dead Node
+    Instruction *FinalNode = dyn_cast<Instruction>(LastAdd);
+    if (!FinalNode)
+      continue;
+
+    // delete the dead PHI Node
+    if (!PN->use_empty())
+      PN->replaceAllUsesWith(PoisonValue::get(PN->getType()));
+    llvm::RecursivelyDeleteDeadPHINode(PN);
```
Comment (arsenm) [inline llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]:
```suggestion
    RecursivelyDeleteDeadPHINode(PN);
```

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 19]
```diff
@@ -0,0 +1,329 @@
...
+
+    if (!RecurrenceInst->use_empty()) {
+      // Copy users to a temporary vector to avoid iterator invalidation
+      SmallVector<User *, 8> Users(RecurrenceInst->user_begin(),
+                                   RecurrenceInst->user_end());
+      for (auto *U : Users) {
+        auto *phi = llvm::dyn_cast<llvm::PHINode>(U);
```
Comment (arsenm) [inline llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]:
```suggestion
        auto *phi = dyn_cast<llvm::PHINode>(U);
```

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 20]
```diff
@@ -0,0 +1,329 @@
...
+
+    // delete the dead PHI Node
+    if (!PN->use_empty())
+      PN->replaceAllUsesWith(PoisonValue::get(PN->getType()));
+    llvm::RecursivelyDeleteDeadPHINode(PN);
+
+    if (!RecurrenceInst->use_empty()) {
+      // Copy users to a temporary vector to avoid iterator invalidation
+      SmallVector<User *, 8> Users(RecurrenceInst->user_begin(),
+                                   RecurrenceInst->user_end());
```
Comment (arsenm) [inline llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: Use early_inc_range?

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp (hunk 16 above)
Comment (Anjian-Wen) [reply llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: Thanks for the review, I will check the comments one by one.

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp (hunk 17 above)
Comment (Anjian-Wen) [reply llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: fixed

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp (hunk 18 above)
Comment (Anjian-Wen) [reply llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: fixed

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp (hunk 19 above)
Comment (Anjian-Wen) [reply llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: fixed

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp (hunk 20 above)
Comment (Anjian-Wen) [reply llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: I have change it to `make_early_inc_range`

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp (hunk 16 above)
Comment (Anjian-Wen) [reply llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: use `any_of` and `make_early_inc_range` to simplify the assign and break pattern.

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 21]
```diff
@@ -0,0 +1,322 @@
...
+  if (!Preheader) {
+    Preheader = InsertPreheaderForLoop(&L, DT, LI, nullptr, false);
+    if (!Preheader) {
+      LLVM_DEBUG(dbgs() << "LRM: Failed to create a preheader for loop "
+                        << Header->getName() << ".\n");
+      return false;
+    }
+  }
+
+  bool transform_success = false;
```
Comment (davemgreen) [inline llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: transform_success -> TransformSuccess or Changed?

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 22]
```diff
@@ -0,0 +1,322 @@
...
+  SmallVector<Instruction *, 8> StackRecur;
+  SmallVector<PHINode *, 8> Stack;
+  int phi_count = 0;
```
Comment (davemgreen) [inline llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: phi_count -> PhiCount etc

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 23]
```diff
@@ -0,0 +1,322 @@
...
+        if (phi && phi->getParent() == LandingPad && !phi->use_empty()) {
+          phi->replaceAllUsesWith(FinalNode);
+          llvm::RecursivelyDeleteDeadPHINode(phi);
+          transform_success = true;
+        }
+      }
+    }
+  }
+
+  if (transform_success) {
```
Comment (davemgreen) [inline llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: return Changed;

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp [hunk 24]
```diff
@@ -0,0 +1,322 @@
...
+      continue;
+
+    Value *RecurrenceValueFromPHI = PN->getIncomingValueForBlock(Latch);
+    Instruction *RecurrenceInst = dyn_cast<Instruction>(RecurrenceValueFromPHI);
+    if (!RecurrenceInst || RecurrenceInst->getNumOperands() != 2)
+      continue;
+
+    // Don't match if the Recurrence Value has other use in loop
+    bool hasOtherUse = llvm::any_of(
+        make_early_inc_range(RecurrenceValueFromPHI->users()), [&](User *U) {
```
Comment (davemgreen) [inline llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: Remove make_early_inc_range here I would think?

**Review by davemgreen — COMMENTED** [review]
Hi - the last time I checked from what I remember this looked good for MVE now - the check for preferInLoopReduction was performing well.  There was still one issue on AArch64 where I think a reduction was being pulled out of an _outer_ loop, and that was just hurting register pressure. I think. But trying it again it was crashing due to the make_early_inc_range so couldn't look into it again. Would you be against limiting this innermost loops to avoid that issue? Hopefully that will help.

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp (hunk 23 above)
Comment (Anjian-Wen) [reply llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: fixed!

File: llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp (hunk 22 above)
Comment (Anjian-Wen) [reply llvm/lib/Transforms/Vectorize/LoopReduceMotion.cpp]: fixed

**Comment by Anjian-Wen** [comment]
@davemgreen Sure, thanks, I have changed the use of make_early_inc_range and passed the tests on my local env. we can look into it to find how to solve the register pressure in Aarch64
